include/klee/Constraints.h

```diff
@@ -12,48 +12,59 @@
 
 #include "klee/Expr.h"
 
-// FIXME: Currently we use ConstraintManager for two things: to pass
-// sets of constraints around, and to optimize constraints. We should
-// move the first usage into a separate data structure
-// (ConstraintSet?) which ConstraintManager could embed if it likes.
 namespace klee {
 
-class ExprVisitor;
-  
-class ConstraintManager {
+/// Resembles a set of constraints that can be passed around
+///
+class ConstraintSet {
+  friend class ConstraintManager;
+
 public:
   typedef std::vector< ref<Expr> > constraints_ty;
   typedef constraints_ty::iterator iterator;
   typedef constraints_ty::const_iterator const_iterator;
-
-  ConstraintManager() {}
-
-  // create from constraints with no optimization
-  explicit ConstraintManager(const std::vector<ref<Expr>> &_constraints);
-
-  typedef std::vector< ref<Expr> >::const_iterator constraint_iterator;
-
-  ref<Expr> simplifyExpr(ref<Expr> e) const;
-
-  void addConstraint(ref<Expr> e);
+  typedef const_iterator constraint_iterator;
 
   bool empty() const;
-  ref<Expr> back() const;
   constraint_iterator begin() const;
   constraint_iterator end() const;
   size_t size() const;
 
-  bool operator==(const ConstraintManager &other) const;
+  explicit ConstraintSet(constraints_ty cs) : constraints(std::move(cs)) {}
+  ConstraintSet() = default;
+
+  void push_back(const ref<Expr> &e);
+
+  bool operator==(const ConstraintSet &b) const {
+    return constraints == b.constraints;
+  }
 
 private:
-  std::vector< ref<Expr> > constraints;
+  constraints_ty constraints;
+};
+
+class ExprVisitor;
+
+/// Manages constraints, e.g. optimisation
+class ConstraintManager {
```

Comment:
I like this renaming to ConstraintSet and more use of static method, but I don't get why keep ConstraintManager around. It seems like the only purpose for it now is to add constraints with this (anti-)pattern:
```
ConstraintManager cm(constraintSet);
cm.addConstraint(e);
```

What is the point? I guess you were trying to separate out the logic for (optionally) simplifying constraints, which I'm not against. But then I would suggestion just having it as a static method, I would also rename the class to `Constraints`, so the above example would be:

```
Constraints::addConstraint(constraintSet, e);
```

but at this point I don't understand why not just keep

```
constraintSet.addConstraint(e);
```

that seems the most natural, but I guess that's just renaming.

Reply:
Think of the `ConstraintManager` as an entity that sometimes need to keep state for multiple invocations, therefore I dediced against a static function.

Reply:
Are there examples currently in the code that need that?

Reply:
In a follow-up PR in preparation there is.

Reply:
In a follow-up PR, there is.